File: 139-word_break.py
# time: O(N*M*len(dict)) ~ cubic
# space: O(n^2)
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        words = set(wordDict) # for faster containment checks
        q = deque([0])
        visited = set()

        while q:
            start = q.popleft()
            # visited.add(start) # optimization needed

            if start == len(s):
                return True

            for end in range(start + 1, len(s) + 1):
                if end in visited: continue

                if s[start:end] in words:
                    q.append(end)
                    visited.add(end) # time optimization for visited

        return False

File: 139-word_break.cs
public class Solution {
    public bool WordBreak(string s, IList<string> wordDict)
    {
        HashSet<string> words = new(wordDict);
        HashSet<int> visited = new();
        Queue<int> q = new();
        q.Enqueue(0);

        while (q.Count() > 0)
        {
            var start = q.Dequeue();

            if (visited.Contains(start))
            {
                continue;
            }

            visited.Add(start);

            if (start == s.Length)
            {
                return true;
            }

            for (var end = start + 1; end < s.Length + 1; end++)
            {
                if (words.Contains(s.Substring(start, end - start)))
                {
                    if (!visited.Contains(end))
                    {
                        q.Enqueue(end);
                    }
                }
            }
        }

        return false;
    }
}
